File: src/test/java/com/rarnu/ison/test/java/TestISONANTIC.java
package com.rarnu.ison.test.java;

import com.rarnu.ison.*;
import kotlin.text.Regex;
import org.junit.Test;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import static org.junit.Assert.*;

public class TestISONANTIC {

    @Test
    public void testStringRequired() {
        var schema = new StringSchema();

        var err = schema.validate("hello");
        assertNull(err);

        err = schema.validate(null);
        assertNotNull(err);
        assert (err.getMessage().contains("required"));
    }

    @Test
    public void testStringOptional() {
        var schema = new StringSchema().optional();

        var err = schema.validate(null);
        assertNull(err);

        err = schema.validate("hello");
        assertNull(err);
    }

    @Test
    public void testStringMinLength() {
        var schema = new StringSchema().min(5);

        var err = schema.validate("hello");
        assertNull(err);

        err = schema.validate("hi");
        assertNotNull(err);
        assert (err.getMessage().contains("at least 5"));
    }

    @Test
    public void testStringMaxLength() {
        var schema = new StringSchema().max(5);

        var err = schema.validate("hello");
        assertNull(err);

        err = schema.validate("hello world");
        assertNotNull(err);
        assert (err.getMessage().contains("at most 5"));
    }

    @Test
    public void testStringExactLength() {
        var schema = new StringSchema().length(5);

        var err = schema.validate("hello");
        assertNull(err);

        err = schema.validate("hi");
        assertNotNull(err);
        assert (err.getMessage().contains("exactly 5"));
    }

    @Test
    public void testStringEmail() {
        var schema = new StringSchema().email();

        var err = schema.validate("test@example.com");
        assertNull(err);

        err = schema.validate("invalid-email");
        assertNotNull(err);
        assert (err.getMessage().contains("invalid email"));
    }

    @Test
    public void testStringURL() {
        var schema = new StringSchema().url();

        var err = schema.validate("https://example.com");
        assertNull(err);

        err = schema.validate("http://example.com/path");
        assertNull(err);

        err = schema.validate("not-a-url");
        assertNotNull(err);
        assert (err.getMessage().contains("invalid URL"));
    }

    @Test
    public void testStringRegex() {
        var pattern = new Regex("^[A-Z]{2,3}$");
        var schema = new StringSchema().regex(pattern);

        var err = schema.validate("AB");
        assertNull(err);

        err = schema.validate("ABC");
        assertNull(err);

        err = schema.validate("A");
        assertNotNull(err);

        err = schema.validate("ABCD");
        assertNotNull(err);
    }

    @Test
    public void testStringDescribe() {
        var schema = new StringSchema().describe("User's name");

        assertEquals("User's name", schema.getDescription());
    }

    @Test
    public void testStringRefine() {
        var schema = new StringSchema().refine(s -> Character.isUpperCase(s.charAt(0)), "must start with uppercase");

        var err = schema.validate("Hello");
        assertNull(err);

        err = schema.validate("hello");
        assertNotNull(err);
        assert (err.getMessage().contains("must start with uppercase"));
    }

    // Number Schema Tests

    @Test
    public void testNumberRequired() {
        var schema = new NumberSchema();

        var err = schema.validate(42.5);
        assertNull(err);

        err = schema.validate(null);
        assertNotNull(err);
    }

    @Test
    public void testNumberOptional() {
        var schema = new NumberSchema().optional();

        var err = schema.validate(null);
        assertNull(err);
    }

    @Test
    public void testIntSchema() {
        var schema = NumberSchema.INT();

        var err = schema.validate(42L);
        assertNull(err);

        err = schema.validate(42.5);
        assertNotNull(err);
        assert (err.getMessage().contains("expected integer"));
    }

    @Test
    public void testNumberMin() {
        var schema = new NumberSchema().min(10.0);

        var err = schema.validate(10.0);
        assertNull(err);

        err = schema.validate(5.0);
        assertNotNull(err);
        assert (err.getMessage().contains("at least"));
    }

    @Test
    public void testNumberMax() {
        var schema = new NumberSchema().max(10.0);

        var err = schema.validate(10.0);
        assertNull(err);

        err = schema.validate(15.0);
        assertNotNull(err);
        assert (err.getMessage().contains("at most"));
    }

    @Test
    public void testNumberPositive() {
        var schema = new NumberSchema().positive();

        var err = schema.validate(5.0);
        assertNull(err);

        err = schema.validate(0.0);
        assertNotNull(err);

        err = schema.validate(-5.0);
        assertNotNull(err);
        assert (err.getMessage().contains("positive"));
    }

    @Test
    public void testNumberNegative() {
        var schema = new NumberSchema().negative();

        var err = schema.validate(-5.0);
        assertNull(err);

        err = schema.validate(0.0);
        assertNotNull(err);

        err = schema.validate(5.0);
        assertNotNull(err);
        assert (err.getMessage().contains("negative"));
    }

    @Test
    public void testNumberRefine() {
        var schema = new NumberSchema().refine(n -> n.intValue() % 2 == 0, "must be even");

        var err = schema.validate(4.0);
        assertNull(err);

        err = schema.validate(3.0);
        assertNotNull(err);
        assert (err.getMessage().contains("must be even"));
    }

    // Boolean Schema Tests

    @Test
    public void testBooleanRequired() {
        var schema = new BooleanSchema();

        var err = schema.validate(true);
        assertNull(err);

        err = schema.validate(false);
        assertNull(err);

        err = schema.validate(null);
        assertNotNull(err);
    }

    @Test
    public void testBooleanOptional() {
        var schema = new BooleanSchema().optional();

        var err = schema.validate(null);
        assertNull(err);
    }

    @Test
    public void testNullSchema() {
        var schema = new NullSchema();

        var err = schema.validate(null);
        assertNull(err);

        err = schema.validate("not null");
        assertNotNull(err);
        assert (err.getMessage().contains("expected null"));
    }

    @Test
    public void testRefRequired() {
        var schema = new RefSchema();

        var err = schema.validate(":1");
        assertNull(err);

        err = schema.validate(Map.of("_ref", "1"));
        assertNull(err);

        err = schema.validate(null);
        assertNotNull(err);
    }

    @Test
    public void testRefOptional() {
        var schema = new RefSchema().optional();

        var err = schema.validate(null);
        assertNull(err);
    }

    @Test
    public void testRefNamespace() {
        var schema = new RefSchema().namespace("user");

        var err = schema.validate(Map.of(
                "_ref", "1",
                "_namespace", "user"
        ));
        assertNull(err);

        err = schema.validate(Map.of(
                "_ref", "1",
                "_namespace", "other"
        ));
        assertNotNull(err);
    }

    @Test
    public void testRefRelationship() {
        var schema = new RefSchema().relationship("OWNS");

        var err = schema.validate(Map.of(
                "_ref", "1",
                "_relationship", "OWNS"
        ));
        assertNull(err);

        err = schema.validate(Map.of(
                "_ref", "1",
                "_relationship", "OTHER"
        ));
        assertNotNull(err);
    }

    @Test
    public void testRefStringFormat() {
        var schema = new RefSchema();

        var err = schema.validate(":1");
        assertNull(err);

        err = schema.validate(":user:42");
        assertNull(err);

        err = schema.validate("not-a-ref");
        assertNotNull(err);
    }

    @Test
    public void testObjectRequired() {
        var schema = ObjectSchema.OBJECT(new HashMap<>(Map.of(
                "name", new StringSchema(),
                "age", NumberSchema.INT()
        )));

        var err = schema.validate(Map.of(
                "name", "Alice",
                "age", 30L
        ));
        assertNull(err);

        err = schema.validate(null);
        assertNotNull(err);
    }

    @Test
    public void testObjectFieldValidation() {
        var schema = ObjectSchema.OBJECT(new HashMap<>(Map.of(
                "name", new StringSchema().min(1),
                "email", new StringSchema().email()
        )));

        var err = schema.validate(Map.of(
                "name", "Alice",
                "email", "alice@example.com"
        ));
        assertNull(err);

        err = schema.validate(Map.of(
                "name", "",
                "email", "invalid"
        ));
        assertNotNull(err);
        var verrs = (ValidationErrors) err;
        assertNotNull(verrs);
        assertEquals(2, verrs.getErrors().size());
    }

    @Test
    public void testObjectOptionalField() {
        var schema = ObjectSchema.OBJECT(new HashMap<>(Map.of(
                "name", new StringSchema(),
                "email", new StringSchema().optional()
        )));

        var err = schema.validate(Map.of(
                "name", "Alice"
        ));
        assertNull(err);
    }

    @Test
    public void testObjectExtend() {
        var baseSchema = ObjectSchema.OBJECT(new HashMap<>(Map.of(
                "id", NumberSchema.INT(),
                "name", new StringSchema()
        )));

        var extendedSchema = baseSchema.extend(Map.of(
                "email", new StringSchema().email()
        ));

        var err = extendedSchema.validate(Map.of(
                "id", 1L,
                "name", "Alice",
                "email", "alice@example.com"
        ));
        assertNull(err);
    }

    @Test
    public void testObjectPick() {
        var schema = ObjectSchema.OBJECT(new HashMap<>(Map.of(
                "id", NumberSchema.INT(),
                "name", new StringSchema(),
                "email", new StringSchema()
        )));

        var pickedSchema = schema.pick("id", "name");

        var err = pickedSchema.validate(Map.of(
                "id", 1L,
                "name", "Alice"
        ));
        assertNull(err);
    }

    @Test
    public void testObjectOmit() {
        var schema = ObjectSchema.OBJECT(new HashMap<>(Map.of(
                "id", NumberSchema.INT(),
                "name", new StringSchema(),
                "email", new StringSchema()
        )));

        var omittedSchema = schema.omit("email");

        var err = omittedSchema.validate(Map.of(
                "id", 1L,
                "name", "Alice"
        ));
        assertNull(err);
    }

    // Array Schema Tests

    @Test
    public void testArrayRequired() {
        var schema = new ArraySchema(new StringSchema());

        var err = schema.validate(List.of("a", "b", "c"));
        assertNull(err);

        err = schema.validate(null);
        assertNotNull(err);
    }

    @Test
    public void testArrayOptional() {
        var schema = new ArraySchema(new StringSchema()).optional();

        var err = schema.validate(null);
        assertNull(err);
    }

    @Test
    public void testArrayMinLength() {
        var schema = new ArraySchema(new StringSchema()).min(2);

        var err = schema.validate(List.of("a", "b"));
        assertNull(err);

        err = schema.validate(List.of("a"));
        assertNotNull(err);
        assert (err.getMessage().contains("at least 2"));
    }

    @Test
    public void testArrayMaxLength() {
        var schema = new ArraySchema(new StringSchema()).max(2);

        var err = schema.validate(List.of("a", "b"));
        assertNull(err);

        err = schema.validate(List.of("a", "b", "c"));
        assertNotNull(err);
        assert (err.getMessage().contains("at most 2"));
    }

    @Test
    public void testArrayItemValidation() {
        var schema = new ArraySchema(NumberSchema.INT());

        var err = schema.validate(List.of(1L, 2L, 3L));
        assertNull(err);

        err = schema.validate(List.of(1L, "not an int", 3L));
        assertNotNull(err);
        var verrs = (ValidationErrors) err;
        assertNotNull(verrs);
        assertEquals(1, verrs.getErrors().size());
        assertEquals("[1]", verrs.getErrors().get(0).getField());
    }

    @Test
    public void testTableRequired() {
        var schema = new TableSchema("users", new HashMap<>(Map.of(
                "id", NumberSchema.INT(),
                "name", new StringSchema()
        )));

        var err = schema.validate(List.of(
                Map.of("id", 1L, "name", "Alice"),
                Map.of("id", 2L, "name", "Bob")
        ));
        assertNull(err);

        err = schema.validate(null);
        assertNotNull(err);
    }

    @Test
    public void testTableOptional() {
        var schema = new TableSchema("users", new HashMap<>(Map.of(
                "id", NumberSchema.INT(),
                "name", new StringSchema()
        ))).optional();

        var err = schema.validate(null);
        assertNull(err);
    }

    @Test
    public void testTableRowValidation() {
        var schema = new TableSchema("users", new HashMap<>(Map.of(
                "id", NumberSchema.INT(),
                "email", new StringSchema().email()
        )));

        var err = schema.validate(List.of(
                Map.of("id", 1L, "email", "alice@example.com"),
                Map.of("id", 2L, "email", "invalid")
        ));
        assertNotNull(err);
        var verrs = (ValidationErrors) err;
        assertNotNull(verrs);
        assertEquals(1, verrs.getErrors().size());
        assert (verrs.getErrors().get(0).getField().contains("row[1]"));
    }

    @Test
    public void testTableBlockFormat() {
        var schema = new TableSchema("users", new HashMap<>(Map.of(
                "id", NumberSchema.INT(),
                "name", new StringSchema()
        )));

        var err = schema.validate(Map.of(
                "kind", "table",
                "name", "users",
                "rows", List.of(
                        Map.of("id", 1L, "name", "Alice")
                )
        ));
        assertNull(err);
    }

    @Test
    public void testTableGetName() {
        var schema = new TableSchema("users", new HashMap<>());
        assertEquals("users", schema.getName());
    }

    @Test
    public void testDocumentParse() {
        var schema = new DocumentSchema(new HashMap<>(Map.of(
                "users", new TableSchema("users", new HashMap<>(Map.of(
                        "id", NumberSchema.INT(),
                        "name", new StringSchema()
                ))),
                "config", ObjectSchema.OBJECT(new HashMap<>(Map.of(
                        "debug", new BooleanSchema()
                )))
        )));

        var doc = Map.of(
                "users", List.of(
                        Map.of("id", 1L, "name", "Alice")
                ),
                "config", Map.of(
                        "debug", true
                )
        );

        var ret = schema.parse(doc);
        assertNotNull(ret.getFirst());
        assertNull(ret.getSecond());
    }

    @Test
    public void testDocumentParseErrors() {
        var schema = new DocumentSchema(new HashMap<>(Map.of(
                "users", new TableSchema("users", new HashMap<>(Map.of(
                        "id", NumberSchema.INT(),
                        "email", new StringSchema().email()
                )))
        )));

        var doc = Map.of(
                "users", List.of(
                        Map.of("id", 1L, "email", "invalid")
                )
        );

        var ret = schema.parse(doc);
        var err = ret.getSecond();
        assertNotNull(err);
        var verrs = (ValidationErrors) err;
        assertNotNull(verrs);
        assertEquals(1, verrs.getErrors().size());
    }

    @Test
    public void testDocumentSafeParse() {
        var schema = new DocumentSchema(new HashMap<>(Map.of(
                "users", new TableSchema("users", new HashMap<>(Map.of(
                        "id", NumberSchema.INT(),
                        "name", new StringSchema()
                )))
        )));

        // Valid document
        var doc = Map.of(
                "users", List.of(
                        Map.of("id", 1L, "name", "Alice")
                )
        );

        var result = schema.safeParse(doc);
        assertTrue(result.getSuccess());
        assertNotNull(result.getData());
        assertNull(result.getError());

        // Invalid document
        var invalidDoc = Map.of(
                "users", List.of(
                        Map.of("id", "not-an-int", "name", "Alice")
                )
        );

        result = schema.safeParse(invalidDoc);
        assertFalse(result.getSuccess());
        assertNull(result.getData());
        assertNotNull(result.getError());
    }

    // I Namespace Tests

    @Test
    public void testINamespace() {
        assertNotNull(I.STRING());
        assertNotNull(I.NUMBER());
        assertNotNull(I.INT());
        assertNotNull(I.FLOAT());
        assertNotNull(I.BOOLEAN());
        assertNotNull(I.BOOL());
        assertNotNull(I.NULL());
        assertNotNull(I.REF());
        assertNotNull(I.REFERENCE());
        assertNotNull(I.OBJECT(new HashMap<>()));
        assertNotNull(I.ARRAY(I.STRING()));
        assertNotNull(I.TABLE("test", new HashMap<>()));
    }

    @Test
    public void testINamespaceUsage() {
        // Example of using I namespace like Zod's z
        var userSchema = I.TABLE("users", Map.of(
                "id", I.INT(),
                "name", I.STRING().min(1),
                "email", I.STRING().email(),
                "active", I.BOOL()
        ));

        var err = userSchema.validate(List.of(
                Map.of(
                        "id", 1L,
                        "name", "Alice",
                        "email", "alice@example.com",
                        "active", true
                )
        ));
        assertNull(err);
    }

    // ValidationError Tests

    @Test
    public void testValidationErrorString() {
        var err = new ValidationError(
                "email",
                "invalid email format",
                "not-an-email"
        );

        assert(err.error().contains("email: invalid email format"));
    }

    @Test
    public void testValidationErrorsString() {
        var errs = new ValidationErrors();
        errs.getErrors().addAll(List.of(
                new ValidationError("email", "invalid email", null),
                new ValidationError("name", "required", null)
        ));

        assert (errs.error().contains("email: invalid email"));
        assert (errs.error().contains("name: required"));
    }

    @Test
    public void testValidationErrorsHasErrors() {
        var empty = new ValidationErrors();
        assertFalse(empty.hasErrors());

        var withErrors = new ValidationErrors();
        withErrors.getErrors().add(new ValidationError("test", "error", null));
        assertTrue(withErrors.hasErrors());
    }

}

File: src/test/kotlin/TestISONANTIC.kt
package com.rarnu.ison.test


import com.rarnu.ison.ArraySchema
import com.rarnu.ison.BooleanSchema
import com.rarnu.ison.DocumentSchema
import com.rarnu.ison.I
import com.rarnu.ison.NullSchema
import com.rarnu.ison.NumberSchema
import com.rarnu.ison.ObjectSchema
import com.rarnu.ison.RefSchema
import com.rarnu.ison.StringSchema
import com.rarnu.ison.TableSchema
import com.rarnu.ison.ValidationError
import com.rarnu.ison.ValidationErrors
import org.junit.Test
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Assert.assertFalse

class TestISONANTIC {
    @Test
    fun testStringRequired() {
        val schema = StringSchema()

        var err = schema.validate("hello")
        assertNull(err)

        err = schema.validate(null)
        assertNotNull(err)
        assert(err?.message?.contains("required") == true)
    }

    @Test
    fun testStringOptional() {
        val schema = StringSchema().optional()

        var err = schema.validate(null)
        assertNull(err)

        err = schema.validate("hello")
        assertNull(err)
    }

    @Test
    fun testStringMinLength() {
        val schema = StringSchema().min(5)

        var err = schema.validate("hello")
        assertNull(err)

        err = schema.validate("hi")
        assertNotNull(err)
        assert(err?.message?.contains("at least 5") == true)
    }

    @Test
    fun testStringMaxLength() {
        val schema = StringSchema().max(5)

        var err = schema.validate("hello")
        assertNull(err)

        err = schema.validate("hello world")
        assertNotNull(err)
        assert(err?.message?.contains("at most 5") == true)
    }

    @Test
    fun testStringExactLength() {
        val schema = StringSchema().length(5)

        var err = schema.validate("hello")
        assertNull(err)

        err = schema.validate("hi")
        assertNotNull(err)
        assert(err?.message?.contains("exactly 5") == true)
    }

    @Test
    fun testStringEmail() {
        val schema = StringSchema().email()

        var err = schema.validate("test@example.com")
        assertNull(err)

        err = schema.validate("invalid-email")
        assertNotNull(err)
        assert(err?.message?.contains("invalid email") == true)
    }

    @Test
    fun testStringURL() {
        val schema = StringSchema().url()

        var err = schema.validate("https://example.com")
        assertNull(err)

        err = schema.validate("http://example.com/path")
        assertNull(err)

        err = schema.validate("not-a-url")
        assertNotNull(err)
        assert(err?.message?.contains("invalid URL") == true)
    }

    @Test
    fun testStringRegex() {
        val pattern = Regex("^[A-Z]{2,3}$")
        val schema = StringSchema().regex(pattern)

        var err = schema.validate("AB")
        assertNull(err)

        err = schema.validate("ABC")
        assertNull(err)

        err = schema.validate("A")
        assertNotNull(err)

        err = schema.validate("ABCD")
        assertNotNull(err)
    }

    @Test
    fun testStringDefault() {
        val schema = StringSchema().default("default")

        val (def, hasDefault) = schema.getDefault()
        assertTrue(hasDefault)
        assertEquals("default", def)
    }

    @Test
    fun testStringDescribe() {
        val schema = StringSchema().describe("User's name")

        assertEquals("User's name", schema.getDescription())
    }

    @Test
    fun testStringRefine() {
        val schema = StringSchema().refine({ s -> s[0].isUpperCase() }, "must start with uppercase")

        var err = schema.validate("Hello")
        assertNull(err)

        err = schema.validate("hello")
        assertNotNull(err)
        assert(err?.message?.contains("must start with uppercase") == true)
    }

    // Number Schema Tests

    @Test
    fun testNumberRequired() {
        val schema = NumberSchema()

        var err = schema.validate(42.5)
        assertNull(err)

        err = schema.validate(null)
        assertNotNull(err)
    }

    @Test
    fun testNumberOptional() {
        val schema = NumberSchema().optional()

        val err = schema.validate(null)
        assertNull(err)
    }

    @Test
    fun testIntSchema() {
        val schema = NumberSchema.INT()

        var err = schema.validate(42L)
        assertNull(err)

        err = schema.validate(42.5)
        assertNotNull(err)
        assert(err?.message?.contains("expected integer") == true)
    }

    @Test
    fun testNumberMin() {
        val schema = NumberSchema().min(10.0)

        var err = schema.validate(10.0)
        assertNull(err)

        err = schema.validate(5.0)
        assertNotNull(err)
        assert(err?.message?.contains("at least") == true)
    }

    @Test
    fun testNumberMax() {
        val schema = NumberSchema().max(10.0)

        var err = schema.validate(10.0)
        assertNull(err)

        err = schema.validate(15.0)
        assertNotNull(err)
        assert(err?.message?.contains("at most") == true)
    }

    @Test
    fun testNumberPositive() {
        val schema = NumberSchema().positive()

        var err = schema.validate(5.0)
        assertNull(err)

        err = schema.validate(0.0)
        assertNotNull(err)

        err = schema.validate(-5.0)
        assertNotNull(err)
        assert(err?.message?.contains("positive") == true)
    }

    @Test
    fun testNumberNegative() {
        val schema = NumberSchema().negative()

        var err = schema.validate(-5.0)
        assertNull(err)

        err = schema.validate(0.0)
        assertNotNull(err)

        err = schema.validate(5.0)
        assertNotNull(err)
        assert(err?.message?.contains("negative") == true)
    }

    @Test
    fun testNumberRefine() {
        val schema = NumberSchema().refine({ n -> n.toInt() % 2 == 0 }, "must be even")

        var err = schema.validate(4.0)
        assertNull(err)

        err = schema.validate(3.0)
        assertNotNull(err)
        assert(err?.message?.contains("must be even") == true)
    }

    // Boolean Schema Tests

    @Test
    fun testBooleanRequired() {
        val schema = BooleanSchema()

        var err = schema.validate(true)
        assertNull(err)

        err = schema.validate(false)
        assertNull(err)

        err = schema.validate(null)
        assertNotNull(err)
    }

    @Test
    fun testBooleanOptional() {
        val schema = BooleanSchema().optional()

        val err = schema.validate(null)
        assertNull(err)
    }

    @Test
    fun testBooleanDefault() {
        val schema = BooleanSchema().default(true)

        val (def, hasDefault) = schema.getDefault()
        assertTrue(hasDefault)
        assertEquals(true, def)
    }

    // Null Schema Tests

    @Test
    fun testNullSchema() {
        val schema = NullSchema()

        var err = schema.validate(null)
        assertNull(err)

        err = schema.validate("not null")
        assertNotNull(err)
        assert(err?.message?.contains("expected null") == true)
    }

    // Reference Schema Tests

    @Test
    fun testRefRequired() {
        val schema = RefSchema()

        var err = schema.validate(":1")
        assertNull(err)

        err = schema.validate(mapOf("_ref" to "1"))
        assertNull(err)

        err = schema.validate(null)
        assertNotNull(err)
    }

    @Test
    fun testRefOptional() {
        val schema = RefSchema().optional()

        val err = schema.validate(null)
        assertNull(err)
    }

    @Test
    fun testRefNamespace() {
        val schema = RefSchema().namespace("user")

        var err = schema.validate(mapOf(
            "_ref" to "1",
            "_namespace" to "user"
        ))
        assertNull(err)

        err = schema.validate(mapOf(
            "_ref" to "1",
            "_namespace" to "other"
        ))
        assertNotNull(err)
    }

    @Test
    fun testRefRelationship() {
        val schema = RefSchema().relationship("OWNS")

        var err = schema.validate(mapOf(
            "_ref" to "1",
            "_relationship" to "OWNS"
        ))
        assertNull(err)

        err = schema.validate(mapOf(
            "_ref" to "1",
            "_relationship" to "OTHER"
        ))
        assertNotNull(err)
    }

    @Test
    fun testRefStringFormat() {
        val schema = RefSchema()

        var err = schema.validate(":1")
        assertNull(err)

        err = schema.validate(":user:42")
        assertNull(err)

        err = schema.validate("not-a-ref")
        assertNotNull(err)
    }

    // Object Schema Tests

    @Test
    fun testObjectRequired() {
        val schema = ObjectSchema.OBJECT(mutableMapOf(
                "name" to StringSchema(),
                "age" to NumberSchema.INT()
            )
        )

        var err = schema.validate(mapOf(
            "name" to "Alice",
            "age" to 30L
        ))
        assertNull(err)

        err = schema.validate(null)
        assertNotNull(err)
    }

    @Test
    fun testObjectFieldValidation() {
        val schema = ObjectSchema.OBJECT(mutableMapOf(
                "name" to StringSchema().min(1),
                "email" to StringSchema().email()
            )
        )

        var err = schema.validate(mapOf(
            "name" to "Alice",
            "email" to "alice@example.com"
        ))
        assertNull(err)

        err = schema.validate(mapOf(
            "name" to "",
            "email" to "invalid"
        ))
        assertNotNull(err)
        val verrs = err as? ValidationErrors
        assertNotNull(verrs)
        assertEquals(2, verrs?.errors?.size)
    }

    @Test
    fun testObjectOptionalField() {
        val schema = ObjectSchema.OBJECT(mutableMapOf(
                "name" to StringSchema(),
                "email" to StringSchema().optional()
            )
        )

        val err = schema.validate(mapOf(
            "name" to "Alice"
        ))
        assertNull(err)
    }

    @Test
    fun testObjectExtend() {
        val baseSchema = ObjectSchema.OBJECT(mutableMapOf(
                "id" to NumberSchema.INT(),
                "name" to StringSchema()
            )
        )

        val extendedSchema = baseSchema.extend(mapOf(
            "email" to StringSchema().email()
        ))

        val err = extendedSchema.validate(mapOf(
            "id" to 1L,
            "name" to "Alice",
            "email" to "alice@example.com"
        ))
        assertNull(err)
    }

    @Test
    fun testObjectPick() {
        val schema = ObjectSchema.OBJECT(mutableMapOf(
                "id" to NumberSchema.INT(),
                "name" to StringSchema(),
                "email" to StringSchema()
            )
        )

        val pickedSchema = schema.pick("id", "name")

        val err = pickedSchema.validate(mapOf(
            "id" to 1L,
            "name" to "Alice"
        ))
        assertNull(err)
    }

    @Test
    fun testObjectOmit() {
        val schema = ObjectSchema.OBJECT(mutableMapOf(
                "id" to NumberSchema.INT(),
                "name" to StringSchema(),
                "email" to StringSchema()
            )
        )

        val omittedSchema = schema.omit("email")

        val err = omittedSchema.validate(mapOf(
            "id" to 1L,
            "name" to "Alice"
        ))
        assertNull(err)
    }

    // Array Schema Tests

    @Test
    fun testArrayRequired() {
        val schema = ArraySchema(StringSchema())

        var err = schema.validate(listOf("a", "b", "c"))
        assertNull(err)

        err = schema.validate(null)
        assertNotNull(err)
    }

    @Test
    fun testArrayOptional() {
        val schema = ArraySchema(StringSchema()).optional()

        val err = schema.validate(null)
        assertNull(err)
    }

    @Test
    fun testArrayMinLength() {
        val schema = ArraySchema(StringSchema()).min(2)

        var err = schema.validate(listOf("a", "b"))
        assertNull(err)

        err = schema.validate(listOf("a"))
        assertNotNull(err)
        assert(err?.message?.contains("at least 2") == true)
    }

    @Test
    fun testArrayMaxLength() {
        val schema = ArraySchema(StringSchema()).max(2)

        var err = schema.validate(listOf("a", "b"))
        assertNull(err)

        err = schema.validate(listOf("a", "b", "c"))
        assertNotNull(err)
        assert(err?.message?.contains("at most 2") == true)
    }

    @Test
    fun testArrayItemValidation() {
        val schema = ArraySchema(NumberSchema.INT())

        var err = schema.validate(listOf(1L, 2L, 3L))
        assertNull(err)

        err = schema.validate(listOf(1L, "not an int", 3L))
        assertNotNull(err)
        val verrs = err as? ValidationErrors
        assertNotNull(verrs)
        assertEquals(1, verrs?.errors?.size)
        assertEquals("[1]", verrs?.errors[0]?.field)
    }

    // Table Schema Tests

    @Test
    fun testTableRequired() {
        val schema = TableSchema("users", mutableMapOf(
            "id" to NumberSchema.INT(),
            "name" to StringSchema()
        ))

        var err = schema.validate(listOf(
            mapOf("id" to 1L, "name" to "Alice"),
            mapOf("id" to 2L, "name" to "Bob")
        ))
        assertNull(err)

        err = schema.validate(null)
        assertNotNull(err)
    }

    @Test
    fun testTableOptional() {
        val schema = TableSchema("users", mutableMapOf(
            "id" to NumberSchema.INT(),
            "name" to StringSchema()
        )).optional()

        val err = schema.validate(null)
        assertNull(err)
    }

    @Test
    fun testTableRowValidation() {
        val schema = TableSchema("users", mutableMapOf(
            "id" to NumberSchema.INT(),
            "email" to StringSchema().email()
        ))

        val err = schema.validate(listOf(
            mapOf("id" to 1L, "email" to "alice@example.com"),
            mapOf("id" to 2L, "email" to "invalid")
        ))
        assertNotNull(err)
        val verrs = err as? ValidationErrors
        assertNotNull(verrs)
        assertEquals(1, verrs?.errors?.size)
        assert(verrs?.errors[0]?.field?.contains("row[1]") == true)
    }

    @Test
    fun testTableBlockFormat() {
        val schema = TableSchema("users", mutableMapOf(
            "id" to NumberSchema.INT(),
            "name" to StringSchema()
        ))

        val err = schema.validate(mapOf(
            "kind" to "table",
            "name" to "users",
            "rows" to listOf(
                mapOf("id" to 1L, "name" to "Alice")
            )
        ))
        assertNull(err)
    }

    @Test
    fun testTableGetName() {
        val schema = TableSchema("users", mutableMapOf())
        assertEquals("users", schema.getName())
    }

    // Document Schema Tests

    @Test
    fun testDocumentParse() {
        val schema = DocumentSchema(mutableMapOf(
            "users" to TableSchema("users", mutableMapOf(
                "id" to NumberSchema.INT(),
                "name" to StringSchema()
            )),
            "config" to ObjectSchema.OBJECT(mutableMapOf(
                "debug" to BooleanSchema()
            ))
        ))

        val doc = mapOf(
            "users" to listOf(
                mapOf("id" to 1L, "name" to "Alice")
            ),
            "config" to mapOf(
                "debug" to true
            )
        )

        val (result, err) = schema.parse(doc)
        assertNull(err)
        assertNotNull(result)
    }

    @Test
    fun testDocumentParseErrors() {
        val schema = DocumentSchema(mutableMapOf(
            "users" to TableSchema("users", mutableMapOf(
                "id" to NumberSchema.INT(),
                "email" to StringSchema().email()
            ))
        ))

        val doc = mapOf(
            "users" to listOf(
                mapOf("id" to 1L, "email" to "invalid")
            )
        )

        val (_, err) = schema.parse(doc)
        assertNotNull(err)
        val verrs = err as? ValidationErrors
        assertNotNull(verrs)
        assertEquals(1, verrs?.errors?.size)
    }

    @Test
    fun testDocumentSafeParse() {
        val schema = DocumentSchema(mutableMapOf(
            "users" to TableSchema("users", mutableMapOf(
                "id" to NumberSchema.INT(),
                "name" to StringSchema()
            ))
        ))

        // Valid document
        val doc = mapOf(
            "users" to listOf(
                mapOf("id" to 1L, "name" to "Alice")
            )
        )

        var result = schema.safeParse(doc)
        assertTrue(result.success)
        assertNotNull(result.data)
        assertNull(result.error)

        // Invalid document
        val invalidDoc = mapOf(
            "users" to listOf(
                mapOf("id" to "not-an-int", "name" to "Alice")
            )
        )

        result = schema.safeParse(invalidDoc)
        assertFalse(result.success)
        assertTrue(result.data == null)
        assertNotNull(result.error)
    }

    // I Namespace Tests

    @Test
    fun testINamespace() {
        assertNotNull(I.STRING())
        assertNotNull(I.NUMBER())
        assertNotNull(I.INT())
        assertNotNull(I.FLOAT())
        assertNotNull(I.BOOLEAN())
        assertNotNull(I.BOOL())
        assertNotNull(I.NULL())
        assertNotNull(I.REF())
        assertNotNull(I.REFERENCE())
        assertNotNull(I.OBJECT(mapOf()))
        assertNotNull(I.ARRAY(I.STRING()))
        assertNotNull(I.TABLE("test", mapOf()))
    }

    @Test
    fun testINamespaceUsage() {
        // Example of using I namespace like Zod's z
        val userSchema = I.TABLE("users", mapOf(
            "id" to I.INT(),
            "name" to I.STRING().min(1),
            "email" to I.STRING().email(),
            "active" to I.BOOL().default(true)
        ))

        val err = userSchema.validate(listOf(
            mapOf(
                "id" to 1L,
                "name" to "Alice",
                "email" to "alice@example.com",
                "active" to true
            )
        ))
        assertNull(err)
    }

    // ValidationError Tests

    @Test
    fun testValidationErrorString() {
        val err = ValidationError(
            field = "email",
            message = "invalid email format",
            value = "not-an-email"
        )

        assertEquals("email: invalid email format", err.error())
    }

    @Test
    fun testValidationErrorsString() {
        val errs = ValidationErrors()
        errs.errors.addAll(listOf(
            ValidationError(field = "email", message = "invalid email", value = null),
            ValidationError(field = "name", message = "required", value = null)
        ))

        assert(errs.error().contains("email: invalid email"))
        assert(errs.error().contains("name: required"))
    }

    @Test
    fun testValidationErrorsHasErrors() {
        val empty = ValidationErrors()
        assertFalse(empty.hasErrors())

        val withErrors = ValidationErrors()
        withErrors.errors.add(ValidationError(field = "test", message = "error", value = null))
        assertTrue(withErrors.hasErrors())
    }
}